File: IOI/IOI2000p1palindrome.py
n = int(input())
w = input()

dp = [[0 for i in range(2)]for i in range(n+1)]

for x in range(1,n+1):
    for y in range(1,n+1):
        if(w[x-1] == w[n-y]):
            dp[y][x%2] = dp[y-1][(x-1)%2] + 1
        else:
            dp[y][x%2] = max(dp[y-1][x%2],dp[y][(x-1)%2])


print(n-dp[n][n%2])
        

File: IOI/IOI2000p1Palindrome.java
package javastuff;
import java.util.*;
import java.lang.Math;
public class IOI2000p1Palindrome {
	static Scanner sc = new Scanner(System.in);

	public static void main(String[] args){
		int n = sc.nextInt();sc.nextLine();
		String word = sc.nextLine();
		int[][] dp = new int[2][n+1];

		for(int x = 1; x <= n; x++){
			for(int y = 1; y <=n; y++){
				if(word.charAt(x-1) == word.charAt(n-y)){
					dp[x%2][y] = dp[(x-1)%2][y] + 1;					
				}
				else{
					dp[x%2][y] = Math.max(dp[x%2][y-1], dp[(x-1)%2][y]);
				}
			}
		}
		System.out.println(n-dp[n%2][n]);
	}
}
